### PR TITLE
Initial PIC implementation

### PR DESCRIPTION
**Pull request by zayfod**
Initial PICmicro MCU implementation:
- test code works
- working `PICAssembler` with tokenizer and instruction encoding
- `PICCentralProcessingUnit` that implements `NOP`, `MOVWF`, `BCF`, `BSF`, `GOTO`, and `MOVLW` instructions

### CODE REVIEW

File: 8_bit_pic.py [hunk 1]
```diff
@@ -11,102 +37,352 @@
 logger = logging.getLogger(__name__)
 
 WORD_SIZE = 8
+FLASH_SIZE = 2 * 1024 * 1024
+RAM_SIZE = 4 * 1024
```
Comment (zayfod) [inline 8_bit_pic.py:41]: Using the same memory space for code and data - 2 MB flash, followed by 4 kB RAM.

File: 8_bit_pic.py [hunk 2]
```diff
@@ -11,102 +37,352 @@
...
+        self.label = label
+
+    def __repr__(self) -> str:
+        return f"Label('{self.label}')"
+
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        assembler.symbol_table["map"][self.label] = len(
+            assembler.byte_code
+        )
+        return b""
```
Comment (zayfod) [inline 8_bit_pic.py:66]: I like the label token idea which helps with populating `map` during instruction encoding.

File: 8_bit_pic.py [hunk 3]
```diff
@@ -11,102 +37,352 @@
...
+        self.d = str(d)
+
+    def __repr__(self) -> str:
+        return f"{self.inst}(f={self.f}, d={self.d})"
+
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        f = resolve_int_value(assembler, self.f, 0x7f)
+        d = resolve_int_value(assembler, self.d, 0x1)
+        code = self.op | (d << 7) | f
+        return code.to_bytes(2, 'big')
```
Comment (zayfod) [inline 8_bit_pic.py:85]: I struggled a bit with the `InstructionMeta` approach as I was looking for a more readable representation of parsed tokens and ended up with this.

File: 8_bit_pic.py [hunk 4]
```diff
@@ -11,102 +37,352 @@
...
 PORTA   equ 05h
+
+        nop
         bsf STATUS,5
         movlw 00h
         movwf TRISA
         bcf STATUS,5
-Start   movlw 02h
+
+Start:  movlw 02h
```
Comment (zayfod) [inline 8_bit_pic.py:179]: Per spec, label are separated by `:`.

File: 8_bit_pic.py [hunk 5]
```diff
@@ -11,102 +37,352 @@
...
+    def __init__(self, k: str) -> None:
+        super().__init__("MOVLW", 0x3000, k, 0xff)
+
+
+TEST_PROG = """
 STATUS  equ 03h
 TRISA   equ 85h
 PORTA   equ 05h
+
+        nop
```
Comment (zayfod) [inline 8_bit_pic.py:173]: Added just for testing.

File: 8_bit_pic.py [hunk 6]
```diff
@@ -11,102 +37,352 @@
...
         movwf TRISA
         bcf STATUS,5
-Start   movlw 02h
+
+Start:  movlw 02h
         movwf PORTA
         movlw 00h
         movwf PORTA
-        goto Star
+        goto Start
```
Comment (zayfod) [inline 8_bit_pic.py:183]: This was a bug :P

File: 8_bit_pic.py [hunk 7]
```diff
@@ -11,102 +37,352 @@
...
         self,
         program_text,
         instruction_codes: type[vm_types.GenericInstructionSet],
         word_size,
     ):
         self.text = program_text
+        self.codes = instruction_codes
+        self.word_size = word_size
+        self.word_size_bytes = word_size // vm_types.BITS_IN_BYTE
+        self.symbol_table = {"map": {}, "symbols": {}}
```
Comment (zayfod) [inline 8_bit_pic.py:202]: I failed to understand what `refs` was for and removed it.

`symbols` is for constants defined with `EQU` or `SET`.

File: 8_bit_pic.py [hunk 8]
```diff
@@ -11,102 +37,352 @@
...
+                # Defines symbol value.
+                if name in self.symbol_table["symbols"]:
+                    raise Exception(f"Symbol '{pseudoop}' redefinition on line {line_num}.")
+                self.symbol_table["symbols"][name] = operands[0]
+            case "SET":
+                # Defines or re-defines symbol value.
+                self.symbol_table["symbols"][name] = operands[0]
+            case "ERROR":
+                # Generates a user-defined error.
+                raise Exception(f"Error {operands[0]}.")
```
Comment (zayfod) [inline 8_bit_pic.py:220]: Added this just for fun.

File: 8_bit_pic.py [hunk 9]
```diff
@@ -11,102 +37,352 @@
...
+            case "BCF":
+                tokens.append(BCFToken(operands[0], operands[1]))
+            case "BSF":
+                tokens.append(BSFToken(operands[0], operands[1]))
+            case "GOTO":
+                tokens.append(GOTOToken(operands[0]))
+            case "MOVLW":
+                tokens.append(MOVLWToken(operands[0]))
+            case _:
+                raise Exception(f"Unexpected opcode '{opcode}' on line {line_num}.")
```
Comment (zayfod) [inline 8_bit_pic.py:241]: I like this static way of dealing with the various tokens as the number and type of operands differs.

File: 8_bit_pic.py [hunk 10]
```diff
@@ -11,102 +37,352 @@
...
+        elif self.ic & 0xf000 == 0x2000:
+            ic = self.ic & 0xf800
+        elif self.ic & 0xf000 == 0x3000:
+            ic = self.ic & 0xff00
+        else:
+            raise Exception(f"Unexpected instruction code 0x{self.ic:02x}.")
+        ic_key = InstructionCodes(ic)
+        logger.debug(f"Executing {ic_key} (0x{self.ic:04x}))...")
+        inst_func = self.INSTRUCTION_MAP[ic_key]
+        inst_func(self, self.ic)
```
Comment (zayfod) [inline 8_bit_pic.py:368]: With PIC MCUs, the opcode location varies and is mixed with arguments :(

File: 8_bit_pic.py [hunk 11]
```diff
@@ -11,102 +37,352 @@
...
-        raise NotImplementedError
+        next_ip = self.pc + self.word_in_bytes
+        self.ic = int.from_bytes(self.RAM[self.pc: next_ip])
+        self.pc = next_ip
 
     def reset(self):
-        return
+        self.ic = 0
+        self.pc = 0
+        self.wreg = 0
```
Comment (zayfod) [inline 8_bit_pic.py:348]: I'll move these to `RAM` eventually and add the status and "top-of-stack registers.

File: 8_bit_pic.py [hunk 12]
```diff
@@ -11,102 +37,352 @@
...
+            case "GOTO":
+                tokens.append(GOTOToken(operands[0]))
+            case "MOVLW":
+                tokens.append(MOVLWToken(operands[0]))
+            case _:
+                raise Exception(f"Unexpected opcode '{opcode}' on line {line_num}.")
+
     def tokenize(self) -> Sequence[vm_types.AssemblerToken]:
-        return []
+        tokens: list[vm_types.AssemblerToken] = []
```
Comment (zayfod) [inline 8_bit_pic.py:244]: The stream of tokens idea is nice.

I guess this method should be a generator?

File: 8_bit_pic.py [hunk 13]
```diff
@@ -11,102 +37,352 @@
...
 
 
 @dataclass
-class CPURegisters: ...
-
+class PICCentralProcessingUnit(vm_types.GenericCentralProcessingUnit):
+    RAM: bytearray
+    INSTRUCTION_MAP: ClassVar[
+        MutableMapping[InstructionCodes, Callable[[Self, int], None]]
+    ] = {}
```
Comment (zayfod) [inline 8_bit_pic.py:320]: This decorator based interface feels clunky.

Some base class with a `decode()` method (think the reverse of `AssemblerToken` could be nice.

**Review by hgueorguiev — APPROVED** [review]

File: 8_bit_pic.py (hunk 13 above)
Comment (hgueorguiev) [reply 8_bit_pic.py:320]: Ya def. working through approaches as I am going ... I did add in a way to register variations of the ins. in my last commit (still decorator based tho) ... which is where I might take this ultimately in the toy vm, maybe 🤔

Comment (izo0x90) [reply 8_bit_pic.py:320]: Ya def. working through approaches as I am going ... I did add in a way to register variations of the ins. in my last commit (still decorator based tho) ... which is where I might take this ultimately in the toy vm, maybe 🤔

File: 8_bit_pic.py (hunk 10 above)
Comment (izo0x90) [reply 8_bit_pic.py:368]: 💀 of course lol

File: 8_bit_pic.py [hunk 14]
```diff
@@ -1,7 +1,33 @@
...
+
+Regerences:
+- https://ww1.microchip.com/downloads/en/DeviceDoc/31029a.pdf
+- https://developerhelp.microchip.com/xwiki/bin/view/products/mcu-mpu/8bit-pic/8-bitbl/8bitblis/
+- https://infonics.wordpress.com/wp-content/uploads/2015/03/introduction-to-pic-18-microcontrollers1.pdf
+- https://en.wikipedia.org/wiki/PIC_instruction_listings#Mid-range_core_devices_(14_bit)
+- https://ww1.microchip.com/downloads/en/DeviceDoc/MPLAB%20XC8%20PIC%20Assembler%20User%27s%20Guide%2050002974A.pdf
+
+"""
+
```
Comment (izo0x90) [inline 8_bit_pic.py:21]: Lmao love this ... Idc why its so exciting to have two arches emulated in a dumb way in a language that has no business implementing this, but it is! 🤣 :feelsgood:

**Review by izo0x90 — APPROVED** [review]

**Pull request merged**